test/Nonlinear/ReverseAD.jl

```diff
@@ -0,0 +1,834 @@
+    MOI.initialize(evaluator, [:Grad, :Jac, :Hess])
+    g = [NaN]
+    x_val = [1.2, 2.3]
+    MOI.eval_constraint(evaluator, g, x_val)
+    @test g == [x_val[1]^2 + x_val[1] * x_val[2] + x_val[2]^2]
+    @test MOI.jacobian_structure(evaluator) == [(1, 1), (1, 2)]
+    J = [NaN, NaN]
+    MOI.eval_constraint_jacobian(evaluator, J, x_val)
+    @test J == [2 * x_val[1] + x_val[2], x_val[1] + 2 * x_val[2]]
+    @test MOI.hessian_lagrangian_structure(evaluator) ==
+          [(1, 1), (2, 2), (2, 1)]
+    H = [NaN, NaN, NaN]
+    MOI.eval_hessian_lagrangian(evaluator, H, x_val, 0.0, [1.5])
+    @test H == 1.5 .* [2.0, 2.0, 1.0]
+    return
+end
+
+function test_constraint_quadratic_multivariate_subexpressions()
+    x = MOI.VariableIndex(1)
+    y = MOI.VariableIndex(2)
+    model = Nonlinear.Model()
+    ex = Nonlinear.add_expression(model, :($x^2))
+    ey = Nonlinear.add_expression(model, :($y^2))
+    exy = Nonlinear.add_expression(model, :($ex + $x * $y))
+    Nonlinear.add_constraint(model, :($exy + $ey), MOI.LessThan(2.0))
+    evaluator =
+        Nonlinear.Evaluator(model, Nonlinear.SparseReverseMode(), [x, y])
+    MOI.initialize(evaluator, [:Grad, :Jac, :Hess])
+    g = [NaN]
+    x_val = [1.2, 2.3]
+    MOI.eval_constraint(evaluator, g, x_val)
+    @test g ≈ [x_val[1]^2 + x_val[1] * x_val[2] + x_val[2]^2]
+    # Jacobian
+    @test MOI.jacobian_structure(evaluator) == [(1, 1), (1, 2)]
+    J = [NaN, NaN]
+    MOI.eval_constraint_jacobian(evaluator, J, x_val)
+    @test J ≈ [2 * x_val[1] + x_val[2], x_val[1] + 2 * x_val[2]]
+    # Jv
+    y, w = [NaN], [1.1, 2.2]
+    MOI.eval_constraint_jacobian_product(evaluator, y, x_val, w)
+    @test y ≈ [(2 * x_val[1] + x_val[2]) (x_val[1] + 2 * x_val[2])] * w
+    # v'J
+    y, w = [NaN, NaN], [1.1]
+    MOI.eval_constraint_jacobian_transpose_product(evaluator, y, x_val, w)
+    wJ = w' * [(2 * x_val[1] + x_val[2]) (x_val[1] + 2 * x_val[2])]
+    @test y ≈ wJ[:]
+    # Hessian-lagrangian
+    @test MOI.hessian_lagrangian_structure(evaluator) ==
+          [(1, 1), (2, 2), (2, 1)]
+    H = [NaN, NaN, NaN]
+    MOI.eval_hessian_lagrangian(evaluator, H, x_val, 0.0, [1.5])
+    @test H ≈ 1.5 .* [2.0, 2.0, 1.0]
+    return
+end
+
+function test_hessian_sparsity_registered_function()
+    x = MOI.VariableIndex(1)
+    y = MOI.VariableIndex(2)
+    z = MOI.VariableIndex(3)
+    f(x, y) = x^2 + y^2
+    function ∇f(g, x...)
+        g[1] = 2x[1]
+        g[2] = 2x[2]
+        return
+    end
+    function ∇²f(H, x...)
+        H[1, 1] = 2
+        H[2, 2] = 2
+        return
+    end
+    model = Nonlinear.Model()
+    Nonlinear.register_operator(model, :f, 2, f, ∇f, ∇²f)
+    Nonlinear.set_objective(model, :(f($x, $z) + $y^2))
+    evaluator =
+        Nonlinear.Evaluator(model, Nonlinear.SparseReverseMode(), [x, y, z])
+    @test_broken :Hess in MOI.features_available(evaluator)
+    # MOI.initialize(evaluator, [:Grad, :Jac, :Hess])
```

Comment:
Add TODOs to uncomment or delete or add an explanation for why these blocks are commented